doc/nextml2015/presentation.tex

```diff
@@ -0,0 +1,1186 @@
+  \end{itemize}
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+from theano import tensor as T
+tensor3 = T.TensorType(
+    broadcastable=(False, False, False),
+    dtype='float32')
+x = T.tensor3()
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Reductions}
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+from theano import tensor as T
+tensor3 = T.TensorType(
+    broadcastable=(False, False, False),
+    dtype='float32')
+x = tensor3()
+total = x.sum()
+marginals = x.sum(axis=(0, 2))
+mx = x.max(axis=1)
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Dimshuffle}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+from theano import tensor as T
+tensor3 = T.TensorType(
+    broadcastable=(False, False, False))
+x = tensor3()
+y = x.dimshuffle((2, 1, 0))
+a = T.matrix()
+b = a.T
+# Same as b
+c = a.dimshuffle((0, 1))
+# Adding to larger tensor
+d = a.dimshuffle((0, 1, 'x'))
+e = a + d
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Indexing}
+  As NumPy!
+  This mean all slices, index selection return view
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+# return views
+a_tensor[int]
+a_tensor[int, int]
+a_tensor[start:stop:step, start:stop:step]
+a_tensor[::-1] # reverse the first dimension
+
+# Advanced indexing, return copy
+a_tensor[an_index_vector] # Supported on GPU.
+a_tensor[an_index_vector, an_index_vector]
+a_tensor[int, an_index_vector]
+a_tensor[an_index_tensor, ...]
+\end{lstlisting}
+\end{frame}
+
+\subsection{Compiling/Running}
+\begin{frame}{Compiling and running expression}
+  \begin{itemize}
+  \item theano.function
+  \item shared variables and updates
+  \item compilation modes
+  \item compilation for GPU
+  \item optimizations
+  \end{itemize}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{theano.function}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+>>> from theano import tensor as T
+>>> x = T.scalar()
+>>> y = T.scalar()
+>>> from theano import function
+>>> # first arg is list of SYMBOLIC inputs
+>>> # second arg is SYMBOLIC output
+>>> f = function([x, y], x + y)
+>>> # Call it with NUMERICAL values
+>>> # Get a NUMERICAL output
+>>> f(1., 2.)
+array(3.0)
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}{Shared variables}
+  \begin{itemize}
+  \item It’s hard to do much with purely functional programming
+  \item ``shared variables'' add just a little bit of imperative programming
+  \item A “shared variable” is a buffer that stores a numerical value for a Theano variable
```

Comment:
Describing a shared variable as a buffer might be confusing to some. It might be more appropriate to describe it as a variable with a state he persists outside of the execution of Theano functions?